service/controller/clusterapi/v29/resource/tcdp/create.go

```diff
@@ -120,22 +156,200 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+				CIDR: cc.Status.ControlPlane.VPC.CIDR,
+			},
+		},
+		TenantCluster: template.ParamsMainSecurityGroupsTenantCluster{
+			VPC: template.ParamsMainSecurityGroupsTenantClusterVPC{
+				ID: "", // TODO
+			},
+		},
+	}
+
+	return securityGroups, nil
+}
+
+func (r *Resource) newSubnets(ctx context.Context, cl v1alpha1.Cluster, md v1alpha1.MachineDeployment) (*template.ParamsMainSubnets, error) {
+	var subnets *template.ParamsMainSubnets
+
+	azs, err := key.StatusAvailabilityZones(md)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	for _, a := range azs {
+		s := template.ParamsMainSubnetsListItem{
+			AvailabilityZone: a.Name,
+			CIDR:             a.Subnet.Public.CIDR, // TODO we need to figure out if this is correct
+			NameSuffix:       strings.ToUpper(a.Name),
+			RouteTableAssociation: template.ParamsMainSubnetsListItemRouteTableAssociation{
+				NameSuffix: strings.ToUpper(a.Name),
+			},
+			TCCP: template.ParamsMainSubnetsListItemTCCP{
+				Subnet: template.ParamsMainSubnetsListItemTCCPSubnet{
+					ID: "", // TODO
+					RouteTable: template.ParamsMainSubnetsListItemTCCPSubnetRouteTable{
+						ID: "", // TODO
+					},
+				},
+				VPC: template.ParamsMainSubnetsListItemTCCPVPC{
+					ID: "", // TODO
+				},
+			},
+		}
+
+		subnets.List = append(subnets.List, s)
+	}
+
+	return subnets, nil
+}
+
+// minDesiredWorkers calculates appropriate minimum value to be set for ASG
+// Desired value and to be used for computation of workerCountRatio.
+//
+// When cluster-autoscaler has scaled cluster and ASG's Desired value is higher
+// than minimum number of instances allowed for that ASG, then it makes sense to
+// consider Desired value as minimum number of running instances for further
+// operational computations.
 //
-//	iamRoles := &template.ParamsMainIAMRoles{
-//		PeerAccessRoleName: key.RolePeerAccess(cr),
-//		Tenant: template.ParamsMainIAMRolesTenant{
-//			AWS: template.ParamsMainIAMRolesTenantAWS{
-//				Account: template.ParamsMainIAMRolesTenantAWSAccount{
-//					ID: cc.Status.TenantCluster.AWSAccountID,
-//				},
-//			},
-//		},
-//	}
+// Example:
+// Initially ASG has minimum of 3 workers and maximum of 10. Due to amount of
+// workload deployed on workers, cluster-autoscaler has scaled current Desired
+// number of instances to 5. Therefore it makes sense to consider 5 as minimum
+// number of nodes also when working on batch updates on ASG instances.
 //
-//	return iamRoles, nil
-//}
+// Example 2:
+// When end user is scaling cluster and adding restrictions to its size, it
+// might be that initial ASG configuration is following:
+// 		- Min: 3
+//		- Max: 10
+// 		- Desired: 10
+//
+// Now end user decides that it must be scaled down so maximum size is decreased
+// to 7. When desired number of instances is temporarily bigger than maximum
+// number of instances, it must be fixed to be maximum number of instances.
+//
+func minDesiredWorkers(minWorkers, maxWorkers, statusDesiredCapacity int) int {
```

Comment:
This is copied from the adapter package. Our goal is to get rid of it over time so I think this is the better place now.